src/quantuminspire/projectq/backend_qx.py

```diff
@@ -58,27 +61,82 @@ def __init__(self, num_runs: int = 1024, verbose: int = 0, quantum_inspire_api:
         self._verbose: int = verbose
         self._cqasm: str = str()
         self._measured_states: Dict[int, float] = {}
         self._measured_ids: List[int] = []
         self._allocation_map: List[Tuple[int, int]] = []
         self._max_qubit_id: int = -1
-        self._full_state_projection = True
         if quantum_inspire_api is None:
             try:
                 quantum_inspire_api = QuantumInspireAPI()
             except AuthenticationError as ex:
                 raise AuthenticationError('Make sure you have saved your token credentials on disk or '
                                           'provide a QuantumInspireAPI instance as parameter to QIBackend') from ex
         self._quantum_inspire_api: QuantumInspireAPI = quantum_inspire_api
-        self._backend_type: Optional[Union[int, str]] = backend_type
-        backend = self._quantum_inspire_api.get_backend_type(backend_type)
-        self._is_simulation_backend = not backend["is_hardware_backend"]
-        self._max_number_of_qubits = backend["number_of_qubits"]
+        self._backend_type: Dict[str, Any] = self._quantum_inspire_api.get_backend_type(backend_type)
+        if num_runs < 1 or num_runs > self._backend_type['max_number_of_shots']:
+            raise ProjectQBackendError(f'Invalid number of runs (num_runs={num_runs})')
+        self._num_runs: int = num_runs
+        self._full_state_projection = not self._backend_type["is_hardware_backend"]
+        self._is_simulation_backend = not self._backend_type["is_hardware_backend"]
+        self._max_number_of_qubits: int = self._backend_type["number_of_qubits"]
+        self._one_qubit_gates: Tuple[Any, ...] = self._get_one_qubit_gates()
+        self._two_qubit_gates: Tuple[Any, ...] = self._get_two_qubit_gates()
+        self._three_qubit_gates: Tuple[Any, ...] = self._get_three_qubit_gates()
+
+    def _get_one_qubit_gates(self) -> Tuple[Any, ...]:
+        if len(self._backend_type['allowed_operations']) > 0:
+            one_qubit_gates = []
+            for keys in self._backend_type['allowed_operations']:
+                if keys in ['single_gates', 'parameterized_single_gates']:
+                    for gate in self._backend_type['allowed_operations'][keys]:
+                        if gate in ['x', 'y', 'z', 'h','s', 'sdag', 't', 'tdag', 'rx', 'ry', 'rz']:
+                            one_qubit_gates += [getattr(sys.modules[__name__], gate.capitalize())]
+        else:
+            one_qubit_gates = [X, Y, Z, H, S, Sdag, T, Tdag, Rx, Ry, Rz]
+        return tuple(one_qubit_gates)
+
+    def _get_two_qubit_gates(self) -> Tuple[Any, ...]:
+        if len(self._backend_type['allowed_operations']) > 0:
+            two_qubit_gates = []
+            for keys in self._backend_type['allowed_operations']:
+                if keys in ['dual_gates', 'parameterized_dual_gates']:
+                    for gate in self._backend_type['allowed_operations'][keys]:
```

Comment:
I would swap the two. Loop over the two gate_sets that we are interested in, and check whether they are present.
```suggestion
        allowed_operations = self._backend_type['allowed_operations']
        if len(allowed_operations) > 0:
            two_qubit_gates = []
            for gate_set in ['dual_gates', 'parameterized_dual_gates']:
                if gate_set in allowed_operations['allowed_operations']:
                    for gate in allowed_operations[gate_set]:
```

Reply:
btw, same for the `_get_one_qubit_gates` and `_get_three_qubit_gates`

Reply:
Done